beacon_node/network/src/network_beacon_processor/rpc_methods.rs

```diff
@@ -320,14 +320,53 @@ impl<T: BeaconChainTypes> NetworkBeaconProcessor<T> {
     pub fn handle_data_columns_by_root_request(
         self: Arc<Self>,
         peer_id: PeerId,
-        _request_id: PeerRequestId,
+        request_id: PeerRequestId,
         request: DataColumnsByRootRequest,
     ) {
-        // TODO(das): implement handler
-        debug!(self.log, "Received DataColumnsByRoot Request";
-            "peer_id" => %peer_id,
-            "count" => request.data_column_ids.len()
+        let mut send_data_column_count = 0;
+
+        for data_column_id in request.data_column_ids.as_slice() {
+            match self.chain.get_data_column_checking_all_caches(
+                data_column_id.block_root,
+                data_column_id.index,
+            ) {
+                Ok(Some(data_column)) => {
+                    send_data_column_count += 1;
+                    self.send_response(
+                        peer_id,
+                        Response::DataColumnsByRoot(Some(data_column)),
+                        request_id,
+                    );
+                }
+                Ok(None) => {} // no-op
+                Err(e) => {
+                    self.send_error_response(
+                        peer_id,
+                        RPCResponseErrorCode::ServerError,
+                        // TODO(das): leak error details to ease debugging
+                        format!("{:?}", e).to_string(),
+                        request_id,
+                    );
+                    error!(self.log, "Error getting data column";
+                        "block_root" => ?data_column_id.block_root,
+                        "peer" => %peer_id,
+                        "error" => ?e
+                    );
+                    return;
+                }
+            }
+        }
+
+        debug!(
+            self.log,
+            "Received DataColumnsByRoot Request";
+            "peer" => %peer_id,
+            "request" => ?request.group_by_ordered_block_root(),
```

Comment:
what is the benefit of ordering by block roots in this log?

Reply:
I introduced `group_by_ordered_block_root` originally to handle the reprocessing of requests. I left it here as it produces a more succint output. The raw request repeats the root for every column. Usually peers requests multiple indexes for the same root, so the output of this is nicer.